packages/router/src/adapters/contract/contract.ts

```diff
@@ -44,6 +45,197 @@ export const removeLiquidityEvt = new Evt<{
+ * @param reason - Reason for the transaction (i.e. transaction type).
+ * @param requestContext - Request context.
+ * @returns TransactionReceipt of the transaction (if it's successful).
+ * @throws Timeout error if the event is not emitted within the timeout period.
+ */
+const waitForRelayer = async (
+  chainId: number,
+  transactionId: string | undefined,
+  evt: Evt<any>,
+  relayerFee: {
+    asset: string;
+    amount: string;
+  },
+  reason: TransactionReason,
+  requestContext: RequestContext,
+) => {
+  const { txService } = getContext();
+  // Listen for event on contract.
+  const { event } = transactionId
+    ? await evt.pipe(({ args }) => args.txData.transactionId === transactionId).waitFor(300_000)
+    : await evt.waitFor(300_000);
+  // Retrieve the receipt.
+  const receipt = await txService.getTransactionReceipt(chainId, event.transactionHash);
+  // If we get a receipt back, then the relayer tx was successful (and the relayer got paid).
+  incrementRelayerFeesPaid(chainId, relayerFee.amount, relayerFee.asset, reason, requestContext);
+  return receipt;
+};
+
+/**
+ * Helper to execute a router contract transaction, first trying relayers (if applicable) and then
+ * resorting to using the router signer with txservice to send if relayers are unavailable.
+ *
+ * @param args.chainId - Chain ID of the transaction.
+ * @param args.routerContractAddress - Address of the router contract.
+ * @param args.encodedData - Encoded data belonging to the transaction.
+ * @param args.params - Params for the transaction's metatx payload.
+ * @param args.signature - Router's signature to relay.
+ * @param args.relayerFee (optional) - Relayer fee info. If undefined, we assume relayers will not be used.
+ * @param args.relayerFee.amount - Relayer fee amount to be paid.
+ * @param args.relayerFee.asset - Relayer fee asset ID.
+ * @param args.reason - Reason for the transaction (i.e. transaction type).
+ * @param methodContext - Method context used for logging.
+ * @param requestContext - Request context used for logging.
+ *
+ * @returns TransactionReceipt of the transaction (if it's successful).
+ * @throws Errors if the trasnsaction failed to execute with relayer network and txservice (our last resort).
+ */
+const sendRouterContractTx = async (
+  args: {
+    chainId: number;
+    routerContractAddress: string;
+    encodedData: string;
+    params: PrepareParams | FulfillParams | CancelParams | RemoveLiquidityParams;
+    signature: string;
+    relayerFee?: {
+      amount: string;
+      asset: string;
+    };
+    reason: TransactionReason;
+  },
+  methodContext: MethodContext,
+  requestContext: RequestContext,
+): Promise<providers.TransactionReceipt> => {
+  const { txService, wallet, logger, messaging } = getContext();
+  const { chainId, routerContractAddress, encodedData, params, signature, relayerFee, reason } = args;
+
+  // Determine which event and payload type to use based on the transaction type.
+  let evt: Evt<any>;
```

Comment:
Why not just make this a generic function that passes in the types rather than typing it as `any`?

Reply:
bc im dumb